engine/frontend/src/component/log/Log.js

```diff
@@ -0,0 +1,178 @@
+import React, { useCallback, useEffect, useRef, useState } from "react";
+import { VariableSizeList as List } from "react-window";
+import { useWindowResize } from "./useWindowResize";
+import { Flex, Box, Spacer } from "@chakra-ui/react";
+import { TriangleDownIcon, DownloadIcon } from '@chakra-ui/icons'
+import parse from 'html-react-parser';
+import hasAnsi from 'has-ansi';
+import stripAnsi from 'strip-ansi';
+import { useClipboard } from '@chakra-ui/react'
+import { saveTextAsFile } from "../../utils/download";
+
+const Convert = require('ansi-to-html');
+const convert = new Convert();
+
+const Row = ({ data, index, setSize, windowWidth}) => {
+  const rowRef = useRef();
+  
+  useEffect(() => {
+    if (rowRef.current !== undefined) {
+      setSize(index, rowRef.current.getBoundingClientRect().height);
+    } 
+  }, [setSize, index, windowWidth]);
+
+  if (data[index] !== undefined && data[index].length !== 0) {
+    let txt = data[index]
+    if (hasAnsi(data[index])) {
+      const parsedAnsi = convert.toHtml(data[index])
+      txt = parse(parsedAnsi)
+    }
+    return (
+        <div
+          ref={rowRef}
+          style={styles.row}
+        >
+         {txt}
+        </div>
+      );
+  } 
+  return null;
+};
+
+export const Log = ({logs, fileName}) => {
+  const { onCopy, value:copyValue, setValue:setCopyValue, hasCopied } = useClipboard("");
+
+  const [initalRender, setInitialRender] = useState(false);
+  const listRef = useRef(null);
+  const sizeMap = useRef({});
+  const testRef = useRef({})
+  const containerRef = useRef({})
+  const [bottomMap, setMaybeBottom] = useState([0, 0])
+  const [previousOffset, setPreviousOffset] = useState(0)
+
+
+  const setSize = useCallback((index, size) => {
+    sizeMap.current = { ...sizeMap.current, [index]: size };
+    listRef.current.resetAfterIndex(index);
+  }, []);
+
+  const getSize = index => sizeMap.current[index] || 50;
+  const [windowWidth, windowHeight] = useWindowResize();
+
+  const [isBottom, setBottom] = useState(true)
+
+  useEffect(() => {
+    const logsWithoutAnsi = logs.map((log)=> {
+      return stripAnsi(log)
+    })
+
+    setCopyValue(logsWithoutAnsi.join("\n"))    
+    // automatically scroll the bottom if it;s at the bottom
+    if (logs.length > 0 && isBottom) {
+      scrollToBottom()
+    }
+
+  }, [logs]);
```

Comment:
@Peeeekay does this mean that on every change to the logs we process the ansi stripping on all the logs? This sounds very labor intensive but let me know if I'm overlooking something. Is there anything we can do to avoid recomputing the same data over and over? Maybe storing with useMemo?

Reply:
maybe add this context in a comment somewhere so it doesn't get forgotten?